### PR TITLE
restrict public interface of training loop

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

All methods that are not part of the public interface get special protection from agent underscore.

- Adds an underscore ( _ ) before pretty much any method and attribute. It looks ugly but you have to get used to it.
- Remove a small piece of dead code
- No other changes

Part of #7938 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/loops/training_epoch_loop.py [hunk 1]
```diff
@@ -41,17 +41,17 @@ def __init__(self, min_steps: int, max_steps: int):
...
+        # the number of batches seen this run, updates immediately after batch_loop.run()
         self.batches_seen: int = 0
-        self.warning_cache: WarningCache = WarningCache()
-        self.epoch_output: Optional[List[List[STEP_OUTPUT]]] = None
 
         self.batch_loop: Optional[TrainingBatchLoop] = None
 
+        self._dataloader_idx: Optional[int] = None
+        self._should_stop: bool = False
+        self._is_last_batch: Optional[bool] = None
```
Comment (carmocca) [inline pytorch_lightning/loops/training_epoch_loop.py:51]: Agent underscore will have to leave this one alone

```python
$ grep -irn "is_last_batch" docs
docs/source/common/optimizers.rst:230:        if self.trainer.is_last_batch and (self.trainer.current_epoch + 1) % n == 0:
docs/source/guides/speed.rst:386:            is_last_batch_to_accumulate = \
docs/source/guides/speed.rst:387:                (batch_idx + 1) % 2 == 0 or self.trainer.is_last_batch
```

Comment (awaelchli) [reply pytorch_lightning/loops/training_epoch_loop.py:51]: oh, we don't actually have a is_last_batch on the Trainer properties hmm. Either have to re-introduce it or defer to `trainer.fit_loop.training_loop.is_last_batch` in docs

Comment (carmocca) [reply pytorch_lightning/loops/training_epoch_loop.py:51]: I would reintroduce it

**Review by carmocca — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by Borda — APPROVED** [review]
